zenoh-jni/src/key_expr.rs

```diff
@@ -152,3 +115,74 @@ pub(crate) unsafe extern "C" fn Java_io_zenoh_jni_JNIKeyExpr_freePtrViaJNI(
 ) {
     Arc::from_raw(ptr);
 }
+
+pub(crate) fn decode_key_expr(env: &mut JNIEnv, key_expr: &JString) -> Result<KeyExpr<'static>> {
+    let key_expr_str = decode_string(env, key_expr).map_err(|err| {
+        Error::Jni(format!(
+            "Unable to get key expression string value: {}",
+            err
+        ))
+    })?;
+    let key_expr = KeyExpr::try_from(key_expr_str).map_err(|err| {
+        Error::Jni(format!(
+            "Unable to create key expression from string: {}",
+            err
+        ))
+    })?;
+    Ok(key_expr)
+}
+
+pub(crate) fn autocanonize_key_expr(
+    env: &mut JNIEnv,
+    key_expr: &JString,
+) -> Result<KeyExpr<'static>> {
+    let key_expr_str = decode_string(env, key_expr).map_err(|err| {
+        Error::Jni(format!(
+            "Unable to get key expression string value: {}",
+            err
+        ))
+    })?;
+    let key_expr = KeyExpr::autocanonize(key_expr_str).map_err(|err| {
+        Error::Jni(format!(
+            "Unable to create key expression from string: {}",
+            err
+        ))
+    })?;
+    Ok(key_expr)
+}
+
+pub(crate) unsafe fn process_key_expr(
+    env: &mut JNIEnv,
+    key_expr_str: &JString,
+    key_expr_ptr: *const KeyExpr<'static>,
+) -> Result<KeyExpr<'static>> {
+    if key_expr_ptr.is_null() {
+        match decode_key_expr(env, key_expr_str) {
+            Ok(key_expr) => Ok(key_expr),
+            Err(err) => Err(Error::Jni(format!(
+                "Unable to process key expression: {}",
+                err
+            ))),
+        }
+    } else {
+        let key_expr = Arc::from_raw(key_expr_ptr);
+        let key_expr_clone = key_expr.deref().clone();
+        std::mem::forget(key_expr);
+        Ok(key_expr_clone)
```

Comment:
I don't quite understand what this is doing. 
is this to get rid of the Arc or Original KeyExpr ? 
but then keep a copy. I need an adult to explain this to me.

Reply:
This is for the case where we provide a valid pointer to a key expression (for cases of KeyExpr instances that were obtained through declaring the key expression from a session for optimization purposes).
In such cases, the pointer is stored in the Kotlin layer as a raw long, which is in reality an Arc pointing to the key expression. We need to keep it valid until the user undeclares the key expression, and we do so by keeping the Arc reference count to the key expression to at least 1. Here doing `std::mem::forget(key_expr)` is meant to stop Rust from collecting the Rust key expr once the function finishes (otherwise the reference count will eventually be 0 and the key expression instance will be cleared).
And since we want to provide the key expression associated to the pointer, we return an arc clone.
Is it clearer?